src/reporter/reporter.py

```diff
@@ -104,7 +104,25 @@ def _validate_payload(payload):
             payload[attr].update({'value': None})
             log().warning(
                 'An entity update is missing value for attribute {}'.format(attr))
+  
 
+def _filter_empty_entities(payload):
+    log().info('Received payload: {}'.format(payload))
+    attrs, length = list(iter_entity_attrs(payload)), (len(payload))
+    Flag = False
+    del_ids = []
+    for j in attrs:
+        key_list = list(payload[j].keys())
+        if 'value' not in payload[j]:
+            return payload
```

Comment:
um, not sure about this :-) what if the entity has an attribute with **no** value field as in `p` below:

    {
        'id': '299531',
        'type': 'AirQualityObserved',
        'p': { 'type': 'string' },
        'pm10': {'type': 'string', 'value': '', 'metadata': {}},
        'pm25': {'type': 'string', 'value': '', 'metadata': {}},
    }

Can you add a test case to check what happens? Thanks!

Reply:
@c0c0n3 , I have checked it for the entity having attribute with no value field as in p, and test case is passing for this case too. Also, i have attached screenshot for the same.Please check
![QL_TestCase](https://user-images.githubusercontent.com/35693908/70426518-2cabda80-1a99-11ea-9921-54ab355b0dde.PNG)
![Capture](https://user-images.githubusercontent.com/35693908/70426730-82808280-1a99-11ea-987f-f13c26f25bbf.PNG)

Reply:
> assert r.status_code == 200

yep, but the problem here is that getting back a `200` does **not** imply the record wasn't added to the DB. Unless I've missed something---which is entirely possible, old age, don't judge me!---the following could actually happen:

1. the reporter gets hit with a payload of e.g. `[ { 'id': '299531', 'type': 'AirQualityObserved', 'p': { 'type': 'string' } } ]`. So there's just one entity which contains only a `p` attribute with no `value` field.
2. that entity gets passed to ` _filter_empty_entities`.
3. the `if` on [line 116](https://github.com/smartsdk/ngsi-timeseries-api/pull/278/files/4dc1bd1b17632cb9100497ddae376bb59482e11b#diff-f796da60daee45177c252e52fc78b4aaR116) evaluates to true, so the entity is passed on to the translator.
4. the translator inserts the entity in the `etairqualityobserved` table successfully.
5. since there's no error, the reporter returns a `200` response to the client.

If you want to test this with an end-to-end/integration test, then you'd have to query the table to make sure the row hasn't been inserted. Just checking the HTTP response code won't cut it. Why not write a unit tests for ` _filter_empty_entities` though, isn't it easier? E.g.

    payload = { 'id': '299531', 'type': 'AirQualityObserved', 'p': { 'type': 'string' } }
    assert _filter_empty_entities(payload) is None

Just an idea.

Reply:
@c0c0n3 , I have updated test case to check entity  wasn't inserted in the DB . If required i will check unit test too.

Reply:
Yea, I saw that, thanks a stack! But like I said earlier what I'm not sure about is what's going to happen if the entity looks like this:

    { 'id': '299531', 'type': 'AirQualityObserved', 'p': { 'type': 'string' } }

which is why I suggested adding a unit test for it. In fact, what I'm still struggling to get a grip on is the below filtering logic:

    for j in attrs:
        key_list = list(payload[j].keys())
        if 'value' not in payload[j]:
            return payload

My (quite possibly wrong!) interpretation is that if an entity has an attribute with no `value` field, then we insert that entity. Is that correct? If so what's the reasoning behind it?

Reply:
@c0c0n3 ,
In  https://github.com/smartsdk/ngsi-timeseries-api/blob/da95f8dd85bca0fcc0a129aa37cd47956b560911/src/reporter/reporter.py#L104 
if it doesn't have string `value` it will update it to None . https://github.com/smartsdk/ngsi-timeseries-api/blob/257a27dcf3927a53005c4244bbee8955eaa785d1/src/reporter/reporter.py#L116
and not insert that entity. :)